lib/src/mt_shared_rss.c

```diff
@@ -0,0 +1,194 @@
+      dbg("%s(%d), rx pkts %u\n", __func__, queue, rx);
+      last_srss_entry = NULL;
+      for (uint16_t i = 0; i < rx; i++) {
+        srss_entry = NULL;
+        hdr = rte_pktmbuf_mtod(pkts[i], struct mt_udp_hdr*);
+        if (hdr->eth.ether_type !=
+            htons(RTE_ETHER_TYPE_IPV4)) { /* non ip, redirect to cni */
+          UPDATE_ENTRY();
+          srss->cni_entry->flow.cb(srss->cni_entry->flow.priv, &pkts[i], 1);
+          continue;
+        }
+        ipv4 = &hdr->ipv4;
+        if (ipv4->next_proto_id != IPPROTO_UDP) { /* non udp, redirect to cni */
+          UPDATE_ENTRY();
+          srss->cni_entry->flow.cb(srss->cni_entry->flow.priv, &pkts[i], 1);
+          continue;
+        }
+        udp = &hdr->udp;
+        MT_TAILQ_FOREACH(srss_entry, &srss->head, next) {
+          if (ipv4->dst_addr == *(uint32_t*)srss_entry->flow.dip_addr &&
+              ntohs(udp->dst_port) == srss_entry->flow.dst_port) { /* match dst ip:port */
+            if (srss_entry != last_srss_entry) UPDATE_ENTRY();
+            srss_pkts[srss_pkts_nb++] = pkts[i];
+            break;
+          }
+        }
+        if (!srss_entry) { /* no match, redirect to cni */
+          UPDATE_ENTRY();
+          srss->cni_entry->flow.cb(srss->cni_entry->flow.priv, &pkts[i], 1);
+        }
+      }
+      if (srss_pkts_nb)
+        last_srss_entry->flow.cb(last_srss_entry->flow.priv, &srss_pkts[0], srss_pkts_nb);
+      rte_pktmbuf_free_bulk(&pkts[0], rx);
+    }
+    pthread_mutex_unlock(&srss->mutex);
+  }
+
+  return 0;
+}
+
+static int srss_tasklet_start(void* priv) { return 0; }
+static int srss_tasklet_stop(void* priv) { return 0; }
+
+struct mt_srss_entry* mt_srss_get(struct mtl_main_impl* impl, enum mtl_port port,
+                                  struct mt_rx_flow* flow) {
+  struct mt_srss_impl* srss = impl->srss[port];
+  struct mt_srss_entry* entry;
+  MT_TAILQ_FOREACH(entry, &srss->head, next) {
+    if ((flow->sys_queue && entry->flow.sys_queue) ||
+        (entry->flow.dst_port == flow->dst_port &&
+         *(uint32_t*)entry->flow.dip_addr == *(uint32_t*)flow->dip_addr)) {
+      return entry;
+    }
+  }
+  entry = mt_rte_zmalloc_socket(sizeof(*entry), mt_socket_id(impl, port));
+  if (!entry) {
+    err("%s(%d), malloc fail\n", __func__, port);
+    return NULL;
+  }
+  entry->flow = *flow;
+  entry->srss = srss;
+  pthread_mutex_lock(&srss->mutex);
+  MT_TAILQ_INSERT_TAIL(&srss->head, entry, next);
+  pthread_mutex_unlock(&srss->mutex);
+  if (flow->sys_queue) srss->cni_entry = entry;
+  return entry;
+}
+
+int mt_srss_put(struct mt_srss_entry* entry) {
+  struct mt_srss_impl* srss = entry->srss;
+
+  pthread_mutex_lock(&srss->mutex);
+  MT_TAILQ_REMOVE(&srss->head, entry, next);
+  pthread_mutex_unlock(&srss->mutex);
+
+  mt_rte_free(entry);
+  return 0;
+}
+
+int mt_srss_init(struct mtl_main_impl* impl) {
+  int num_ports = mt_num_ports(impl);
+
+  for (int i = 0; i < num_ports; i++) {
+    if (!mt_has_rss(impl, i)) continue;
```

Comment:
Should be mt_has_srss? For other non-shared rss mode, it should leave impl->srss[i] to zero.